include/boost/compute/algorithm/copy.hpp

```diff
@@ -114,29 +307,266 @@ dispatch_copy_async(InputIterator first,
-        return copy_to_host(first, last, result, queue);
+    typedef typename std::iterator_traits<OutputIterator>::value_type output_type;
+    typedef typename InputIterator::value_type input_type;
+
+    const context &context = queue.get_context();
+    const device &device = queue.get_device();
+
+    // loading parameters
+    std::string cache_key =
+        std::string("__boost_compute_copy_to_host_")
+            + type_name<input_type>() + "_" + type_name<output_type>();
+    boost::shared_ptr<parameter_cache> parameters =
+        detail::parameter_cache::get_global_cache(device);
+
+    size_t map_copy_threshold;
+    size_t direct_copy_threshold;
+
+    // calculate default values of thresholds
+    if (device.type() & device::gpu) {
+        // GPUs
+        map_copy_threshold = 524288;  // 0.5 MB
+        direct_copy_threshold = 52428800; // 50 MB
     }
     else {
-        // for non-contiguous input we first copy the values to
-        // a temporary std::vector and then copy from there
-        typedef typename std::iterator_traits<InputIterator>::value_type T;
-        std::vector<T> vector(iterator_range_size(first, last));
+        // CPUs and other devices
+        map_copy_threshold = 134217728; // 128 MB
+        direct_copy_threshold = 0; // it's never efficient for CPUs
+    }
+
+    // load thresholds
+    map_copy_threshold =
+        parameters->get(
+            cache_key, "map_copy_threshold", map_copy_threshold
+        );
+    direct_copy_threshold =
+        parameters->get(
+            cache_key, "direct_copy_threshold", direct_copy_threshold
+        );
+
+    // select copy method based on thresholds & input_size_bytes
+    size_t count = iterator_range_size(first, last);
+    size_t input_size_bytes = count * sizeof(input_type);
+
+    // [0; map_copy_threshold) -> copy_to_host_map()
+    if(input_size_bytes < map_copy_threshold) {
+        return copy_to_host_map(first, last, result, queue);
+    }
+    // [map_copy_threshold; direct_copy_threshold) -> copy [first;last) to
+    //     temporary vector then copy (and convert) to result using std::copy()
+    else if(input_size_bytes < direct_copy_threshold) {
+        std::vector<input_type> vector(count);
         copy_to_host(first, last, vector.begin(), queue);
         return std::copy(vector.begin(), vector.end(), result);
     }
+
+    // [direct_copy_threshold; inf) -> map [result; result + input_size) to
+    //     device and run copy kernel on device for copying & casting
+    // map host memory to device
+    buffer mapped_host(
+        context,
+        count * sizeof(output_type),
+        buffer::write_only | buffer::use_host_ptr,
+        static_cast<void*>(
+            ::boost::addressof(*result)
+        )
+    );
+    copy_on_device(
+        first,
+        last,
+        make_buffer_iterator<output_type>(mapped_host),
+        queue
+    );
+    // update host memory asynchronously by maping and unmaping memory
```

Comment:
Shouldn't `asynchronously` here be `synchronously`? Also, I think an alternative would be instead of mapping and unmapping the buffer, replace the call above with `copy_on_device_async()` and wait on the returned event/future.